### DISCUSSION TITLE
[Contrib help] Seperating the borders

### DISCUSSION THREAD

**Discussion by LamprosPitsillos** [qtile/qtile#3633] [Ideas]
I am interested in trying out separating the borders.
What i mean by that is,
All 4 borders have independent:
- *Colors*
- *Width*

I would like to start with the width first.

I have found what i think is the start of where i should be looking at,
```py
# windows.py
    def paint_borders(self, depth, colors, borderwidth, width, height):
```
But i would like some more guidance on the structure of the project according to this kind of stuff.
Any help from more experienced Contribs would be welcomed!

* inspiration 
https://www.reddit.com/r/unixporn/comments/vcphbh/berry_pine/
![rzgqwjp7tq591](https://user-images.githubusercontent.com/61395246/174067051-e590885b-287b-4f5a-b01c-be730b3d3979.png)

**Comment by m-col** [comment]
Wow that is a nice rice, I can see why you'd want that!

Is this for X11 or Wayland?

**Reply by LamprosPitsillos** [reply]
> Wow that is a nice rice, I can see why you'd want that!

I know right!? 
I am pretty sure that the rounded corners would be really difficult to implement (if not impossible , without a compositor anyway ) but that is a problem for another day.

https://berrywm.org/

Not sure but i think berry is for X11 only

**Reply by m-col** [reply]
Are you trying to get this working for x11 or wayland?

**Reply by LamprosPitsillos** [reply]
X11
I am not aiming at replicating this. I just wanted to see if a could get something similar ( or just something new ) by separating the border attributes to its 4 sides and thought about contributing if i could make it work.

The screenshot was just where i got the idea for this

**Reply by LamprosPitsillos** [reply]
The implementation would look similar to the one for margins ,if a list is given is used for its separate sides, if a single value is given it's a homogeneous border.

**Reply by m-col** [reply]
Cool. The approach for x11 and wayland would be very different internally (and much more cumbersome for x11 unfortunately). This was very briefly discussed here https://github.com/qtile/qtile/issues/1699 but in general I think this is something people would welcome in qtile if implemented neatly. Different features that you describe above would need different things so I'll describe them separately.

### Colours

In x11 windows usually only have a border, which is an equal sized space around the window It can either be given a pixel (i.e. single color that paints the whole thing) or a pixmap which is a pixel buffer the same size of the window including border-width. In Qtile, both of these are already implemented. The former method is used if a user wants a single border color: https://github.com/qtile/qtile/blob/master/libqtile/backend/x11/window.py#L434. If a user wants multiple colours, then the `with PixmapID(self.conn.conn) as pixmap` block a bit further down draws rectangles of decreasing size (drawing multiple borders outside-in) onto a pixmap and then applies that pixmap to the window's border pixmap with `_set_borderpixmap`. In theory you can draw whatever you want onto that pixmap and it will get drawn, and all you'd need to do is modify the way `paint_borders` draws borders. I played around with this idea here which may be a start if you wanted to play with that: https://mcol.xyz/2020/05/hacking-on-qtile-painting-complex-borders.html

### Widths

Things get more complicated if you also want to have differently sized border widths and title bars and such. I don't know much about it but you'd need to make the X11 backend reparent windows so that windows are essentially embedded within a qtile-controlled window that exists solely for drawing window decorations on. Your best bet for figuring that out is another reparenting wm, like berrywm or i3.

**Reply by LamprosPitsillos** [reply]
I must admit I am overwhelmed.. But I was always looking for an excuse to learn more about X11/Wayland .
I will try studying the source code , and give it a shot!

It might sound dump but what resources would you recommend other than the ones mentioned in the docs?
I have to start from somewhere 😄

**Reply by m-col** [reply]
Unfortunately docs in X are pretty poor. IMO the best resources are the sources of different WMs. Small, simple WMs that have one key feature, e.g. berry here with its border shapes, are good places to look as generally the code is straight forward and easy to understand.

**Reply by elParaguayo** [reply]
This is something I might look at it in the future (but as a learning exercise rather than ever wanting to implement it).

It's also not something that could be easily incorporated into qtile-extras.

**Reply by LamprosPitsillos** [reply]
> but as a learning exercise rather than ever wanting to implement it

That's mostly the reason i wanted to try it 😁

**Reply by elParaguayo** [reply]
Well, if you get stuck, let me know and we may be able to learn together!
